### ISSUE TITLE
Why does Number#times return the number?

### ISSUE THREAD

**Issue by TiZ-EX1** [andrewplummer/Sugar#444]
Let's say I want to make an array of five random numbers. Having used underscore, my first instinct will be to do this:

```
var arr = (5).times(function() { return Number.random(50, 100); });
```

But this won't work; it'll uselessly kick back 5. So this method lets you do something five times, and then continue chaining the number. I'm not sure why anyone would want to do that. If you want to preserve this behavior, why not return an array of the returned values if the function it calls returns something?

**Comment by andrewplummer** [comment]
Sorry for the delay on this one. @lolmaus is correct that this is more of a use case for an array constructor, however I agree that the return value for `times` is much more useful as an array as well. Sugar now polyfills `Array.from` which is in the ES6 spec so you can use this and I would argue that it is more semantically correct:

```
Array.from(new Array(5), Math.random);
```

`Number#times` will now produce an identical result:

```
(5).times(Math.random);
```

Personally, I would use `Array.from` as I feel it is clearer if your goal is to create an array. However it's up to you.

**Comment by vendethiel** [comment]
Means it's a perfoirmance hit to use `Number#times` when you don't want the result though, right?

**Comment by andrewplummer** [comment]
`times` is very simple. It's simply pushing to an array and returning, so any performance hit would be trivially minor.

**Comment by vendethiel** [comment]
If you do "500.times", it means you'll allocate a 500-items array... Doesn't sound that great

**Comment by andrewplummer** [comment]
Sounds like a job for a for loop then

**Comment by andrewplummer** [comment]
Well look, it's not that I don't see your point. It's just that I know how easy it is to fall into the trap of premature optimizations. Sugar could be much more performant if we just stripped out most of the features. Of course you're right though, 500 array items isn't ideal, nor is 10,000, nor is 100,000. I think we need to think of the main use case here. In my mind it was something a lot lower, but thinking about it that's an assumption. People certainly may use this to execute a function a large number of times, and if they're not concerned about the result then creating unneeded arrays is a waste.

What I'd like to avoid here are feature flags for the sake of performance. The less burden when using a method, the better. So here's an idea... what if we don't push a result if it's undefined? That way if people are not returning anything in their callback then it will only be an array of 1?

**Comment by vendethiel** [comment]
the check would probably be slow enough that it wouldn't matter. Let's leave it that way :-).

**Comment by andrewplummer** [comment]
Sorry I'm not sure what you mean?

**Comment by vendethiel** [comment]
we'd need to check the return value of each call before adding it to the array. Actually, that might be interesting, but just as a "filter" kind of method...

**Comment by andrewplummer** [comment]
Ok time to be clear. First of all, pushing a value to an array is not directly a "performance hit" in any case in the normal sense. It is a fast operation. However you are correct that creating large objects in memory when you don't need them is not good. It makes more work for the GC and will indirectly affect the performance of your app. This is why I proposed a check for undefined so that callbacks that do not provide a return value.

> the check would probably be slow enough that it wouldn't matter

Why would you think that a simple check for undefined would have [any affect on performance whatsoever](http://jsperf.com/simple-undefined-check), let alone relative to the operation that's going to be executed in `times`? Now we really are into premature optimization territory -- well into it. If you have an operation that is *so* performance sensitive that a simple undefined check will have any affect on your application -- aside from questioning what you're doing in your app -- I would *absolutely* recommend a for loop for you.

**Comment by vendethiel** [comment]
> First of all, pushing a value to an array is not directly a "performance hit" in any case in the normal sense. It is a fast operation.

Well, if you initialized the array with `new Array(i);`

> Why would you think that a simple check for undefined would have any affect on performance whatsoever

Because I know anything like your test case is just optimized away, and you're just benchmarking an empty loop. The VM are smart enough to drop both the if (trivially true) and the Math.pow (inlined, then removed because the value is not used, or the other way around).

This is trivial to demonstrate, by adding an empty loop to your jsbin case: http://jsperf.com/simple-undefined-check/2

On my computer, "without undefined check" is faster than the empty loop (which is the slowest).

**Comment by vendethiel** [comment]
Also, I actually like the "push unless undefined" idea. Seems like semantics I'd use in a real-world app.

**Comment by andrewplummer** [comment]
> Well, if you initialized the array with new Array(i);

Well ok, but that would defeat the purpose. Definitely would push to an empty array.

> Because I know anything like your test case is just optimized away

[OK, fair enough](http://jsperf.com/simple-undefined-check/3). I dunno, this may be optimized somehow as well. That's partially the point though. Just trying to set up a test like this puts us in JS quantum land. It's so trivial that it's hard to know if results are even accurate.

**Comment by vendethiel** [comment]
>  I dunno, this may be optimized somehow as well. 

Yeah, the compiler is able to do LICM (Loop invariant code motion) and trivially prove that `x` will not change during the loop. It can thus remove it. And even if that didn't work, the compiler knows that `Math.pow` has no side-effect, and thus can be removed

**Comment by andrewplummer** [comment]
> and trivially prove that x will not change during the loop. It can thus remove it.

Ok, updated so that it's now in the test case itself, not "outside the clocked loop"

> And even if that didn't work, the compiler knows that Math.pow has no side-effect, and thus can be removed

I wonder if that can be proven... any idea to set up a more meaningful test case then?

> Also, I actually like the "push unless undefined" idea.

Well we can just sit here arguing over nothing then :)

**Comment by andrewplummer** [comment]
Okok... So I guess if we're going to get into this better to do it as realistically as possible, so here we go:

http://jsperf.com/sugar-times-with-result

This is my best attempt. In the end the function sent to `times` is simply updating a global variable, so this may well be optimized away as well (if you have any other ideas I'm all ears). Sugar runs in node too so I didn't want to touch a DOM node or anything browser specific.

**Comment by vendethiel** [comment]
No idea, this is just voodoo :). Maybe it should be logged or something... but that'd mean we'd perf Array#toString and console.log, which is very suboptimal. But I'm happy with these semantics.

**Comment by andrewplummer** [comment]
> No idea, this is just voodoo

It really is, isn't it :)

So if you're happy with the semantics and the perf looks OK (I'm assuming you're seeing the same results I am), then you're ok with this change?

**Comment by lolmaus** [comment]
All I want from `n.times(callback)` is a non-humiliating alternative to `for (let i = 0; i < n; i++) { callback(i) }`.

I definitely don't want a CPU and memory penalty.

If I wanted to assemble an array, I would rather use `Array.from(new Array(5), Math.random)` (or, even better, a [convenience method](https://github.com/andrewplummer/Sugar/issues/471) like `Array.construct(5, Math.random)`.

Looks like `.times()` IS the convenience array construction method I had been asking for. But why the heck it's in the `Number.prototype` space and not `Array`? And if `.times()` is used for arrays constructing, where is the normal `.times()` -- fast, predictable and returning an [iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols)?

Anyone who comes from Ruby (and those guys are your primary audience) would definitely not expect how `.times()` is implemented.

TL/DR:

1. Refactor the current implementation of `Number.prototype.times(callback)` to `Arryay.construct(number, callback)`.
2. Implement a fast `.times()` that does not consume CPU/memory and returns an [iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols).

**Comment by andrewplummer** [comment]
> All I want from n.times(callback) is a non-humiliating alternative to for (let i = 0; i < n; i++) { callback(i) }.

Unfortunately for the world, your use case is not necessarily everyone else's. Also, that's exactly what times does.

> I definitely don't want a CPU and memory penalty.

Avoiding that is exactly what is being discussed right now.

> If I wanted to assemble an array, I would rather use Array.from(new Array(5), Math.random) (or, even better, a convenience method like Array.construct(5, Math.random).

And that's exactly what is going to be supported.

> Looks like .times() IS the convenience array construction method I had been asking for.

No. `Array.from` is.

> But why the heck it's in the Number.prototype space and not Array? 

Because the number of iterations is a number. Not everyone wants to create an array, even if we're discussing allowing that as an option.

> Anyone who comes from Ruby (and those guys are your primary audience)

No idea where you got that idea. Javascript is the primary audience.

> Refactor the current implementation of Number.prototype.times(callback) to Arryay.construct(number, callback).

This will not happen. `Array.from` is made for your use case. Use it.

> Implement a fast .times() that does not consume CPU/memory and returns an iterator.

This most definitely will not happen. Sugar needs to be backward compatible and ES6 just isn't there yet. This would be the most unexpected of all the scenarios discussed so far.

**Comment by andrewplummer** [comment]
I'd like to think about this in real use cases. Let's say you have a like button that instead of 1 like sends out 5 ... Okok maybe not that realistic, but in any case having a single event kick off multiple asynchronous operations shouldn't be that much of a stretch. In this case, you may want to collect promise objects or some other return value that results from the operation. I would argue that this is not a job for an array method as constructing an array is not the point here, but the operation itself is. The semantics look better and the method reads better, and I don't think it's a stretch at all to want to support this.

But I'd really prefer to hear from @TiZ-EX1 about his original use case. What were you trying to accomplish originally?

@lolmaus I think that for you, the point is when your use case *is* constructing arrays, why should it be easier to do that using a syntax that makes less semantic sense for you? I think it's a fair point, and I'm willing to consider an `Array.construct` as a convenience method for your use case as well.

**Comment by TiZ-EX1** [comment]
I hate to say this, but it's been so long since I opened this that I don't even remember what I wanted to do with it aside from the example case I initially posted. Sorry. :(

**Comment by andrewplummer** [comment]
Well that's definitely my fault for taking so long, so I'm sorry as well... Your example does seem to work better with `Array.construct` as @lolmaus suggests, so if it was something like that, then that should cover you. However I think it's work talking through the use cases of `times` as well...

**Comment by lolmaus** [comment]
If you just want to execute something a lot times, array construction would be a useless CPU/memory burden.

But I should say that I've looked into lodash, and their `_.times()` also builds an array.

**Comment by andrewplummer** [comment]
OK this one is now locked down. `Array.construct` will be supported, so @lolmaus gets his syntax. Additionally, not only will `times` only collect non-undefined return values of the callback, it won't even initialize the array until the first defined value is received, so the performance hit is exactly that of an `=== undefined` check (or in other words nothing).
